### DISCUSSION TITLE
Input file in shortcodes

### DISCUSSION THREAD

**Discussion by Nenuial** [quarto-dev/quarto-cli#2230] [Q&A]
Is there a way of knowing the name of the input file (.qmd) in the Lua script of a shortcode?

**Comment by mcanouil** [comment]
Yes, `PANDOC_STATE`, see <https://pandoc.org/lua-filters.html#global-variables>.
It contains `PANDOC_STATE.input_files` (<https://pandoc.org/lua-filters.html#type-commonstate>).

**Reply by cderv** [reply]
I believe this variable will contain the path to the file pass to pandoc which will be a `.md` file. You can probably derive the name of the `qmd` file from it though

I have open https://github.com/quarto-dev/quarto-cli/issues/2249 to track the idea of adding a new utility function for that. 

@Nenuial you could add you use case there and add a 👍

**Reply by Nenuial** [reply]
I tried this, but I'm afraid the original filename gets lost along the different quarto steps. If I access `PANDOC_STATE.input_files` all I get is something along the lines of `/var/folders/gd/rrsyq__147x_8y_3zcpymj_40000gn/T/quarto-session662eb8d7/d63bc63f/quarto-input2e60596b.md`.

@cderv I will do that! It would be great to have something along those lines. I'll explain my use case in the issue.

**Reply by mcanouil** [reply]
There are other fields in `PANDOC_STATE` which might contain something more useful, i.e., resource_path, source_url, output_file.

**Reply by Nenuial** [reply]
Thanks! I had looked at a few of them, but didn't try `output_file` which does contain what I need.

**Reply by cscheid** [reply]
@Nenuial I've found it useful to run `quarto.utils.dump` on these objects to get an idea of everything that is in them. It can be a bit overwhelming to read through, but it helps to remove the "guess the field" aspect from the problem.

**Reply by Nenuial** [reply]
Indeed, that was my first approach. But dumping `PANDOC_STATE` only yielded an unhelpful `pandoc CommonState: 0x6000011a4f78`. I then dumped a few of the variables listed on the pandoc website, but I didn't try the `output_file` one.

**Reply by cscheid** [reply]
Oof, that's really horrible. We need to improve that routine to work better with this data.

**Reply by Nenuial** [reply]
The `quarto.utils.dump` utility is great!! It makes things so much easier than when I was trying to make Lua filters for RMarkdown. If you can make it *compatible* with even more types that would be really appreciated.

**Reply by cscheid** [reply]
See #2254, this will be merged shortly. Sorry about that!

**Reply by Nenuial** [reply]
Thank you, this is great!

**Comment by hamelsmu** [comment]
How do you get the input file, exactly?   when I dump the `PANDOC_STATE`, I get the following output.  I guess I want to get the relative path of the file (relative to `_quarto.yml`)

```
output_file: docs.html
user_data_dir: /Applications/quarto/share/pandoc/datadir
input_files: table: 0x6000008e6d80
  1: /var/folders/hz/fg03h5c13x96l0g9420_qvlm0000gn/T/quarto-session558f13e0/fdee4908/quarto-inputb46c113f.md
request_headers: table: 0x6000008e6d80
  <empty table>
trace: false
verbosity: ERROR
log: table: 0x6000008e6d80
  1: pandoc LogMessage: 0x6000008e6eb8
  2: pandoc LogMessage: 0x6000008e7278
  3: pandoc LogMessage: 0x6000008e72b8
  4: pandoc LogMessage: 0x6000008e7338
  5: pandoc LogMessage: 0x6000008e7378
resource_path: table: 0x6000008e6d80
  1: .
resource_path: table: 0x600000885500
  1: .
output_file: docs.out.ipynb
request_headers: table: 0x600000885500
  <empty table>
log: table: 0x600000885500
  <empty table>
trace: false
input_files: table: 0x600000885500
  1: /var/folders/hz/fg03h5c13x96l0g9420_qvlm0000gn/T/quarto-session558f13e0/fdee4908/quarto-input7ba14add.md
verbosity: ERROR
user_data_dir: /Applications/quarto/share/pandoc/datadir
```

The `input_files` field seems to refer to a temporary file of some kind?   What am I missing in this situation?
